### PR TITLE
Python3 compatible version of txt2tags

### PR DESCRIPTION
**Pull request by nishimotz**
Issue: nvaccess/nvda#8734

Related PR: nvaccess/nvda#9648

It replaces txt2tags.py with my work of https://github.com/nvdajp/txt2tags.

### CODE REVIEW

File: python/txt2tags.py [hunk 1]
```diff
@@ -1506,6 +1507,7 @@ def getRegexes():
 	  'login' : r'A-Za-z0-9_.-',             # for ftp://login@domain.com
 	  'pass'  : r'[^ @]*',                   # for ftp://login:pass@dom.com
 	  'chars' : r'A-Za-z0-9%._/~:,=$@&+-',   # %20(space), :80(port), D&D
+	  'echar' : r'A-Za-z0-9%_/~:,=$@&+-',    # endmost char (does not include period)
```
Comment (LeonarddeR) [inline python/txt2tags.py:1510]: This line, for example

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Are you sure you used the right base version of txt2tags/ There seem to be some lines with unnecessary/unexpected changes in this.

File: python/txt2tags.py (hunk 1 above)
Comment (nishimotz) [reply python/txt2tags.py:1510]: 'end of the word' behaves differently in Python 3, so I have modified like that.

https://github.com/nvdajp/txt2tags/commit/60db7e3572da7230d191f7c922bce2b4b894a4c9

File: python/txt2tags.py [hunk 2]
```diff
@@ -1516,9 +1518,9 @@ def getRegexes():
 	
 	# [ http:// ] [ username:password@ ] domain.com [ / ]
 	#     [ #anchor | ?form=data ]
-	retxt_url = r'\b(%s%s|%s)[%s]+\b/*(\?[%s]+)?(#[%s]*)?'%(
+	retxt_url = r'\b(%s%s|%s)[%s]+[%s]/*(\?[%s]+)?(#[%s]*)?'%(
```
Comment (LeonarddeR) [inline python/txt2tags.py:1521]: Why this change?

File: python/txt2tags.py [hunk 3]
```diff
@@ -1581,11 +1583,11 @@ def getUnknownErrorMessage():
 def Message(msg,level):
 	if level <= VERBOSE and not QUIET:
 		prefix = '-'*5
-		print "%s %s"%(prefix*level, msg)
+		print("%s %s"%(prefix*level, msg))
 def Debug(msg,id=0,linenr=None):
 	"Show debug messages, categorized (colored or not)"
 	if QUIET or not DEBUG: return
-	if int(id) not in range(8): id = 0
+	if int(id) not in list(range(8)): id = 0
```
Comment (LeonarddeR) [inline python/txt2tags.py]: Wrapping this into a list doesn't seem to be necessary

File: python/txt2tags.py [hunk 4]
```diff
@@ -1604,25 +1606,34 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
 	else:
-		try: f = open(file); data = f.readlines() ; f.close()
+		try:
+			f = open(file, encoding='utf-8')
+			data = f.readlines()
+			f.close()
 		except:
 			if not ignore_error:
 				Error(_("Cannot read file:")+" %s"%file)
+		# remove UTF-8 BOM
```
Comment (LeonarddeR) [inline python/txt2tags.py]: Why would there be an UTF-8 BOM in a file that is opened as utf-8? Shouldn't that already discard the BOM from the data?

File: python/txt2tags.py [hunk 5]
```diff
@@ -1604,25 +1606,34 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
+		if data and data[0] and data[0][0] == '\ufeff':
+			data[0] = data[0][1:]
 	if remove_linebreaks:
-		data = map(lambda x:re.sub('[\n\r]+$','',x), data)
+		data = [re.sub('[\n\r]+$','',x) for x in data]
 	Message(_("File read (%d lines): %s")%(len(data),file),2)
 	return data
 def Savefile(file, contents):
 	try: f = open(file, 'wb')
```
Comment (LeonarddeR) [inline python/txt2tags.py:1621]: Shouldn't you open the file as utf-8, or doesn't that make any difference?

File: python/txt2tags.py [hunk 6]
```diff
@@ -1604,25 +1606,34 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
 	except: Error(_("Cannot open file for writing:")+" %s"%file)
-	if type(contents) == type([]): doit = f.writelines
+	if type(contents) == type([]):
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
	if isinstance(contents, list):
```

File: python/txt2tags.py [hunk 7]
```diff
@@ -1604,25 +1606,34 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
+		doit = f.write
+		contents = ('\n'.join([s.rstrip('\n') for s in contents]) + '\n').encode('utf-8')
+	elif type(contents) == type([]): doit = f.writelines
```
Comment (LeonarddeR) [inline python/txt2tags.py]: What's wrong with writelines? You  now have to write the line breaks yourself, and haven't used the \r character.

File: python/txt2tags.py [hunk 8]
```diff
@@ -1826,7 +1837,7 @@ def compose_cmdline(self, conf={}, no_check=0):
 			# Default values are useless on cmdline
 			if val == dft_options.get(key): continue
 			# -short format
-			if key in use_short.keys():
+			if key in list(use_short.keys()):
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
			if key in use_short:
```

File: python/txt2tags.py [hunk 9]
```diff
@@ -2101,25 +2112,25 @@ def add(self, key, val):
 			key = key[3:]              # remove prefix
 			val = self.off.get(key)    # turn key OFF
 		# Is this key valid?
-		if key not in self.defaults.keys():
+		if key not in list(self.defaults.keys()):
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
		if key not in self.defaults:
```

File: python/txt2tags.py [hunk 10]
```diff
@@ -2101,25 +2112,25 @@ def add(self, key, val):
...
-			if self.parsed.has_key(key):
+			if key in self.parsed:
 				del self.parsed[key]
 			# Nothing more to do
 			return
 		# Flags ON comes empty. we'll add the 1 value now
 		if val == '' and (
-		   key in self.dft_flags.keys() or
-		   key in self.dft_actions.keys()):
+		   key in list(self.dft_flags.keys()) or
```
Comment (LeonarddeR) [inline python/txt2tags.py]: Same as above

File: python/txt2tags.py [hunk 11]
```diff
@@ -2206,8 +2217,8 @@ def parse(self):
 		raw = self.get_target_raw()
 		for target, key, value in raw:
 			self.add(key, value)
-		Message(_("Added the following keys: %s")%string.join(
-		         self.parsed.keys(),', '),2)
+		Message(_("Added the following keys: %s")% ', '.join(
+		         list(self.parsed.keys())),2)
```
Comment (LeonarddeR) [inline python/txt2tags.py]: Makes sense to join the keys here, but you do not need to wrap that in a list I believe. Join will simply iterate over the keys

File: python/txt2tags.py [hunk 12]
```diff
@@ -3327,36 +3336,36 @@ def dumpConfig(source_raw, parsed_config):
...
+	print(_('Full PARSED config'))
+	keys = list(parsed_config.keys()) ; keys.sort()  # sorted
 	for key in keys:
 		val = parsed_config[key]
 		# Filters are the last
 		if key == 'preproc' or key == 'postproc':
 			continue
 		# Flag beautifier
-		if key in FLAGS.keys() or key in ACTIONS.keys():
+		if key in list(FLAGS.keys()) or key in list(ACTIONS.keys()):
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
		if key in FLAGS or key in ACTIONS:
```

File: python/txt2tags.py [hunk 13]
```diff
@@ -4033,7 +4041,7 @@ def get_include_contents(file, path=''):
 	id = 't2t'
 	# Set include type and remove identifier marks
 	mark = file[0]
-	if mark in ids.keys():
+	if mark in list(ids.keys()):
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
	if mark in ids:
```

File: python/txt2tags.py [hunk 14]
```diff
@@ -4565,14 +4573,14 @@ def convert(bodylines, config, firstlinenr=1):
 def load_GUI_resources():
```
Comment (LeonarddeR) [inline python/txt2tags.py:4568]: How difficult would it be to strip the gui from this?

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Wow, this work must have been huge.

**Comment by nishimotz** [comment]
resolved some conversations which are addressed and confirmed with the test environment of the nvdajp repository.

File: python/txt2tags.py [hunk 15]
```diff
@@ -2124,8 +2124,8 @@ def add(self, key, val):
 			return
 		# Flags ON comes empty. we'll add the 1 value now
 		if val == '' and (
-		   key in list(self.dft_flags.keys()) or
-		   key in list(self.dft_actions.keys())):
+		   key in self.dft_flags or
+		   key in dft_actions):
```
Comment (LeonarddeR) [inline python/txt2tags.py]: please restore the removal of self.

Comment (nishimotz) [reply python/txt2tags.py]: oops, sorry.

**Comment by LeonarddeR** [comment]
Do you need any clarification for the open conversations?

**Comment by nishimotz** [comment]
I will address remaining things later.

File: python/txt2tags.py (hunk 4 above)
Comment (nishimotz) [reply python/txt2tags.py]: deleted in favor of open(file, encoding='utf-8-sig')

File: python/txt2tags.py (hunk 5 above)
Comment (nishimotz) [reply python/txt2tags.py:1621]: Actually
```
data = [x.rstrip() for x in data]
```
is mostly the same, however, but not exactly the same.
If a line in the t2t file has trailing spaces, the current code does not remove them.
I wanted to produce exactly the same output, otherwise, it is hard to know the code has regression or not.

File: python/txt2tags.py (hunk 2 above)
Comment (nishimotz) [reply python/txt2tags.py:1521]: It actually uses the 'echar,' which is introduced above.

File: python/txt2tags.py (hunk 14 above)
Comment (nishimotz) [reply python/txt2tags.py:4568]: Rather than stripping the gui, I have worked further regarding gui.
So far, just
```
python3.7 txt2tags.py --gui
```
works, but the actual job is not tested.

File: python/txt2tags.py (hunk 5 above)
Comment (LeonarddeR) [reply python/txt2tags.py:1621]: I don't understand how your remark about `data = [x.rstrip() for x in data]` relates to opening the file with utf-8

Comment (nishimotz) [reply python/txt2tags.py:1621]: I understand you are saying about open(file, 'wb').
I remember that binary mode was necessary for producing the same outputs as with python 2 version, but I should try a better approach.

https://github.com/nvdajp/txt2tags/commit/87808f4833b631d1ff3bc396b01af5bbb858aabd

Comment (nishimotz) [reply python/txt2tags.py:1621]: I have looked at the code again. However, I am reluctant to change further.
SaveFile() is called with the content which is the return value of addLineBreaks().
addLineBreaks() replaces line breaks in condideration of sys.platform.
I don't want to allow platform dependent line ending handling again in SaveFile().

Comment (nishimotz) [reply python/txt2tags.py:1621]: Removed unnecessary string replaces.

File: python/txt2tags.py (hunk 14 above)
Comment (LeonarddeR) [reply python/txt2tags.py:4568]: For NVDA, we really don't care about the gui, but I think it is ok.

File: python/txt2tags.py [hunk 16]
```diff
@@ -1604,25 +1606,30 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
+		data = [re.sub('[\n\r]+$','',x) for x in data]
 	Message(_("File read (%d lines): %s")%(len(data),file),2)
 	return data
 def Savefile(file, contents):
 	try: f = open(file, 'wb')
 	except: Error(_("Cannot open file for writing:")+" %s"%file)
-	if type(contents) == type([]): doit = f.writelines
+	if isinstance(contents, list):
+		doit = f.write
+		contents = (''.join(contents)).encode('utf-8')
```
Comment (LeonarddeR) [inline python/txt2tags.py]: I think I prefer:

```suggestion
		contents = [line.encode('utf-8') for line in contents]
```

And then keep using writelines. This is how it always has been

**Review by LeonarddeR — COMMENTED** [review]
I agree with writing in binary mode, that makes sense.

File: python/txt2tags.py [hunk 17]
```diff
@@ -1604,25 +1606,30 @@ def Readfile(file, remove_linebreaks=0, ignore_error=0):
...
-	if type(contents) == type([]): doit = f.writelines
+	if isinstance(contents, list):
+		doit = f.writelines
+		contents = [line.encode('utf-8') for line in contents]
 	else: doit = f.write
 	doit(contents) ; f.close()
 
 def showdic(dic):
-	for k in dic.keys(): print "%15s : %s" % (k,dic[k])
+	for k in dic.keys(): print("%15s : %s" % (k,dic[k]))
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
	for k, v in dic.items():
		print("%15s : %s" % (k,l))
```

File: python/txt2tags.py [hunk 18]
```diff
@@ -1715,18 +1722,18 @@ def _compose_short_opts(self):
 		ret = []
 		for opt in self.short_long.keys():
 			long = self.short_long[opt]
-			if long in self.all_options: # is flag or option?
+			if int in self.all_options: # is flag or option?
```
Comment (LeonarddeR) [inline python/txt2tags.py]: This looks like a false positive from 2to3. long really should still be long here.

File: python/txt2tags.py [hunk 19]
```diff
@@ -3327,36 +3332,36 @@ def dumpConfig(source_raw, parsed_config):
...
 			val    = val or _('ON')
-			print '  %-8s %s: %s'%(target,key,val)
-		print
+			print('  %-8s %s: %s'%(target,key,val))
+		print()
 	# Then the parsed results of all of them
-	print _('Full PARSED config')
-	keys = parsed_config.keys() ; keys.sort()  # sorted
+	print(_('Full PARSED config'))
+	keys = list(parsed_config.keys()) ; keys.sort()  # sorted
```
Comment (LeonarddeR) [inline python/txt2tags.py]:
```suggestion
	keys = sorted(parsed_config.keys())
```

File: python/txt2tags.py [hunk 20]
```diff
@@ -3327,36 +3332,36 @@ def dumpConfig(source_raw, parsed_config):
...
 		val = parsed_config[key]
 		# Filters are the last
 		if key == 'preproc' or key == 'postproc':
 			continue
 		# Flag beautifier
-		if key in FLAGS.keys() or key in ACTIONS.keys():
+		if key in FLAGS or key in ACTIONS:
 			val = onoff.get(val) or val
 		# List beautifier
 		if type(val) == type([]):
```
Comment (LeonarddeR) [inline python/txt2tags.py]: While this is still valid code, I don't think it is very good coding style. Could you change all these cases as follows?

```suggestion
		if isinstance(val, list):
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Almost there

File: python/txt2tags.py (hunk 17 above)
Comment (nishimotz) [reply python/txt2tags.py]: actually, showdic() is not used anywhere.

**Review by LeonarddeR — APPROVED** [review]

**Review by michaelDCurran — APPROVED** [review]
